File: fs/readfile.js
const fs = require("fs")

/*
const start = process.hrtime();
const rs = fs.createReadStream("./large_file_js.txt");

let count = 0;
rs.on("data", (buf) => {
  count += Buffer.byteLength(buf)

});

rs.on("end", () => {
  console.log("total count: ", count);
  const diff = process.hrtime(start);
  console.log("time consumes: ", Math.floor(diff[0] * 1e3 + diff[1] / 1e6));
});
*/

const BUF_SIZE = 4096 * 3;
function main() {
  const buf = Buffer.alloc(BUF_SIZE);
  try {
    const start = process.hrtime();
    const fd = fs.openSync("./large_file_js.txt", "r");
    let count = 0;
    for(;;) {
      try {
        const bytesRead = fs.readSync(fd, buf, 0, BUF_SIZE);
        if (bytesRead == 0) {
          console.log("read file ends");
          break;
        }
        count += bytesRead;
      } catch (err) {
        console.log(err);
        break;
      }
    }
    console.log("total count: ", count);
    const diff = process.hrtime(start);
    console.log("time consumes: ", Math.floor(diff[0] * 1e3 + diff[1] / 1e6));
  } catch (err) {
    console.log(err);
  }
}

main();

File: fs/readfile.go
package main

import (
  "bufio"
  "os"
  "time"
  "fmt"
  "io"
  "strconv"
)

func main() {
  argc := len(os.Args)
  if argc < 2 || argc > 3 {
    fmt.Println("Usage: go run readfile.go foo.txt 4096")
    return
  }
  buf_size := 4096
  if argc == 3 {
    if s, err := strconv.Atoi(os.Args[2]); err == nil {
      buf_size = s
    }
  }
  start := time.Now()
  var buf = make([]byte, buf_size)
  f, err := os.Open(os.Args[1])
  if err != nil {
    fmt.Println(err)
    return
  }
  r := bufio.NewReader(f)
  count := 0
  for {
    c, err := r.Read(buf)
    if err != nil {
      if err == io.EOF {
        break
      }
      fmt.Println(err)
      return
    }
    count += c
  }
  fmt.Println("total file size: ", count)

  fmt.Println("time consumes: ", time.Now().Sub(start))
}


